### PR TITLE
Diagnostics: Adds total number of active clients information

### LINKED ISSUE #2589: CosmosDiagnostics: Add active client count
The diagnostics currently only has the number of created clients. This issue is to track adding the number of active clients. This way we can see if a user is properly disposing of clients for certain cases.

### PR DESCRIPTION
**Pull request by sourabh1007**
## Description

To Identify if a user is properly disposing of clients for certain cases., Adding `NumberOfActiveClients` information in request Diagnostics.

image

## Type of change

- [] New feature (non-breaking change which adds functionality)

## Closing issues

To automatically close an issue: closes #2589

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 1]
```diff
@@ -262,6 +265,8 @@ protected CosmosClient()
             clientOptions ??= new CosmosClientOptions();
 
             this.ClientId = this.IncrementNumberOfClientsCreated();
+            
+            this.IncrementNumberOfActiveClientsCreated();
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs:269]: Why not reuse the existing method and just add it to IncrementNumberOfClientsCreated?

File: Microsoft.Azure.Cosmos/src/Diagnostics/CosmosTraceDiagnostics.cs [hunk 2]
```diff
@@ -54,6 +54,11 @@ internal bool IsGoneExceptionHit()
             return this.WalkTraceTreeForGoneException(this.Value);
         }
 
+        internal int ActiveClients()
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Diagnostics/CosmosTraceDiagnostics.cs:57]: Why is this needed?

File: Microsoft.Azure.Cosmos/src/Tracing/TraceData/ClientConfigurationTraceDatum.cs [hunk 3]
```diff
@@ -46,6 +47,7 @@ public ReadOnlyMemory<byte> SerializedJson
         {
             get
             {
+                this.cachedNumberOfActiveClient = CosmosClient.NumberOfActiveClientsCreated;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Tracing/TraceData/ClientConfigurationTraceDatum.cs:50]: Why isn't this following the same logic as cachedNumberOfClientCreated?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/DefaultTracingTests.cs [hunk 4]
```diff
@@ -19,8 +19,9 @@ public class DefaultTracingTests
         [TestMethod]
         public void DefaultTracingEnableTest()
         {
-            // Access cosmos client to cause the static consturctor to get called
+            // Access cosmos client to cause the static constructor to get called
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/DefaultTracingTests.cs:22]: Please add specific test to verify the count is correct and is properly decremented after a client is disposed of.

File: Microsoft.Azure.Cosmos/src/Diagnostics/CosmosTraceDiagnostics.cs (hunk 2 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Diagnostics/CosmosTraceDiagnostics.cs]: We need this information for Client telemetry so exposed a function here in order to use it. But I think you are right we really don't need this. We can have this directly.

File: Microsoft.Azure.Cosmos/src/Tracing/TraceData/ClientConfigurationTraceDatum.cs (hunk 3 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Tracing/TraceData/ClientConfigurationTraceDatum.cs:50]: It is not part of UserAgent string and Number Of Active client can be changed without a change in UserAgent.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/DefaultTracingTests.cs (hunk 4 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/DefaultTracingTests.cs:22]: tests are added

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 5]
```diff
@@ -102,6 +102,9 @@ public class CosmosClient : IDisposable
         private bool isDisposed = false;
 
         internal static int numberOfClientsCreated;
+
+        internal static int NumberOfActiveClientsCreated;
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs:106]: Should it be named NumberActiveClients? The Created is a bit redundant?

File: Microsoft.Azure.Cosmos/src/Tracing/TraceData/ClientConfigurationTraceDatum.cs (hunk 3 above)
Comment (asketagarwal) [reply Microsoft.Azure.Cosmos/src/Tracing/TraceData/ClientConfigurationTraceDatum.cs:50]: This will need to be added to the check below that the activeclient count hasn't changed, otherwise the value in the diagnostics won't be correct

File: Microsoft.Azure.Cosmos/src/Tracing/TraceData/ClientConfigurationTraceDatum.cs [hunk 6]
```diff
@@ -46,6 +47,7 @@ public ReadOnlyMemory<byte> SerializedJson
...
                 if ((this.cachedUserAgentString != this.UserAgentContainer.UserAgent) ||
                     (this.cachedNumberOfClientCreated != CosmosClient.numberOfClientsCreated))
```
Comment (asketagarwal) [inline Microsoft.Azure.Cosmos/src/Tracing/TraceData/ClientConfigurationTraceDatum.cs:52]: This will need to be added to the check here that the activeclient count hasn't changed, otherwise the value in the diagnostics won't be correct

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 7]
```diff
@@ -1287,6 +1303,7 @@ protected virtual void Dispose(bool disposing)
                 if (disposing)
                 {
                     this.ClientContext.Dispose();
+                    this.DecrementNumberOfActiveClientsCreated();
```
Comment (asketagarwal) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs:1306]: Possibility of a race condition here ? Could lead to inconsistent results. Maybe make the Dispose() method thread safe ?

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs (hunk 5 above)
Comment (asketagarwal) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs:106]: nit: static variables starting with small case letters ?

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs (hunk 7 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs:1306]: `Interlocked.Decrement(ref NumberOfActiveClients)` is Thread Safe

Comment (asketagarwal) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs]: I am talking about the Dispose() method here. It can be called by multiple threads and there can be a race condition where multiple threads decrement the counter

**Comment by kirankumarkolli** [comment]
PR description is stale, please refresh.

**Review by kirankumarkolli — APPROVED** [review]

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/ActiveClientDiagnosticTest.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    public class ActiveClientDiagnosticTest
+    {
+        private const string ConnectionString = "AccountEndpoint=https://localtestcosmos.documents.azure.com:443/;AccountKey=425Mcv8CXQqzRNCgFNjIhT424GK99CKJvASowTnq15Vt8LeahXTcN5wt3342vQ==;";
+
+        private int prevActiveClients = 0;
+
+        [TestInitialize]
+        public void Initialize()
+        {
+            this.prevActiveClients = CosmosClient.NumberOfActiveClients;
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/ActiveClientDiagnosticTest.cs:22]: Why is this needed? You could just simply set it to 0, right?

```suggestion
            CosmosClient.NumberOfActiveClients = 0;
```

**Review by j82w — DISMISSED** [review]

File: Microsoft.Azure.Cosmos/src/Tracing/TraceData/ClientConfigurationTraceDatum.cs [hunk 9]
```diff
@@ -79,6 +84,8 @@ private ReadOnlyMemory<byte> GetSerializedDatum()
 
             jsonTextWriter.WriteFieldName("NumberOfClientsCreated");
             jsonTextWriter.WriteNumber64Value(this.cachedNumberOfClientCreated);
+            jsonTextWriter.WriteFieldName("NumberOfActiveClients");
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Tracing/TraceData/ClientConfigurationTraceDatum.cs:87]: nit: PR description needs to be updated.

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 10]
```diff
@@ -1280,17 +1302,22 @@ public void Dispose()
...
                 }
-
                 this.isDisposed = true;
             }
+
+            this.DisposedDateTimeUtc = DateTime.UtcNow;
+
+            if (disposing)
+            {
+                this.ClientContext.Dispose();
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs:1318]: Shouldn't this be inside the lock? Otherwise it will Dispose it twice if multiple threads are calling Dispose(true)

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 11]
```diff
@@ -1280,17 +1302,22 @@ public void Dispose()
...
+                this.DecrementNumberOfActiveClients();
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs:1319]: Please add a test where multiple parallel Tasks are calling dispose and check the number of ActiveClients matches, ideally if 2 or more concurrent Tasks call Dispose on the same client, the decrement should be called once

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs (hunk 10 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs:1318]: Nevermind, you are exiting early on the new flow

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs (hunk 11 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs:1319]: That will ensure that the logic you added with the lock is working as expected.

**Review by j82w — APPROVED** [review]
